### DISCUSSION TITLE
How to create an elliptical mask ?

### DISCUSSION THREAD

**Discussion by Ruchika-K** [libvips/libvips#2971] [Q&A]
I want to create an elliptical mask similar to what we do in opencv -
getSructuringElement(https://docs.opencv.org/3.4/d4/d86/group__imgproc__filter.html#gac342a1bb6eabf6f55c803b09268e36dc)  with morph shape as ellipse.
So given a height and width how to create an elliptical mask?

**Comment by jcupitt** [comment]
Hi again, you could make a circle I guess and squash it, would that work? Or do you need an slanted elipse?

You can make a circular mask with `mask_ideal`:

```
$ vips mask_ideal x.csv 9 9 1 --optical --reject
$ more x.csv
0	0	0	0	1	0	0	0	0
0	0	1	1	1	1	1	0	0
0	1	1	1	1	1	1	1	0
0	1	1	1	1	1	1	1	0
1	1	1	1	1	1	1	1	1
0	1	1	1	1	1	1	1	0
0	1	1	1	1	1	1	1	0
0	0	1	1	1	1	1	0	0
0	0	0	0	1	0	0	0	0
```

**Reply by Ruchika-K** [reply]
Thanks @jcupitt ! I don't need a slant ellipse. 
Have a few follow up ques:-
1.How to select the frequency cutoff? 
2. What do you mean by squashing it?
3. I am not getting the expected output by doing this  - 
```
$ vips mask_ideal x.csv 11 7 1 --optical --reject
$ cat x.csv
0	0	0	0	0	1	0	0	0	0	0
0	0	1	1	1	1	1	1	1	0	0
0	1	1	1	1	1	1	1	1	1	0
1	1	1	1	1	1	1	1	1	1	1
0	1	1	1	1	1	1	1	1	1	0
0	0	1	1	1	1	1	1	1	0	0
0	0	0	0	0	1	0	0	0	0	0
```
Expected output (creating an elliptical mask in opencv) is :- 

```
>>> cv2.getStructuringElement(cv2.MORPH_ELLIPSE,(11,7))
array([[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0],
       [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0]], dtype=uint8)
```
4. What if I want a rectangular mask? Does mask_ideal by default create a circular mask?

**Reply by jcupitt** [reply]
Oh, looks like it will generate an elliptical mask directly, that's good. Frequency cutoff is the radius, from 0 to 1 (using the filter meaning).

You can use it from python, of course:

```
$ python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> pyvips.Image.mask_ideal(11, 9, 1, reject=True, optical=True)
<pyvips.Image 11x9 float, 1 bands, fourier>
>>> pyvips.Image.mask_ideal(11, 9, 1, reject=True, optical=True).numpy()
array([[0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 1., 1., 1., 1., 1., 1., 1., 0., 0.],
       [0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
       [0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
       [0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
       [0., 0., 1., 1., 1., 1., 1., 1., 1., 0., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.]], dtype=float32)
```

Is the opencv mask correct? It seems to be missing the leftmost and rightmost columns.

Yes, `mask_ideal` makes an idea pass or reject filter. There are other filter builders, eg. `mask_butterworth` etc. If you just want a simple square of 1s use eg.:

```
>>> (pyvips.Image.black(11, 9) + 1).numpy()
array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]], dtype=float32)
```

**Reply by Ruchika-K** [reply]
Yes the opencv mask I have mentioned above is what we need. We have been using opencv and now for the same functionality we are going to use govips. Hence we need exactly similar results.  I am not able to understand how to replicate the same mask in libvips/govips.

**Reply by jcupitt** [reply]
I doubt if you'll get identical results between any two image processing systems for any non-trivial task. Operations do not have exact specifications and most systems will implement them slightly differently.

You can't even be sure of identical results with the same version of the same image processing system complied in an identical way. What if you run on a CPU without a vector unit? You'll fall back to a C path and that'll almost certainly behave differently. What if the platform swaps libm for libcrm? Again, you'll get different results.

I would set a target result and a tolerance, and probably quite a generous tolerance.

In this case, the opencv mask looks slightly incorrect (to me, perhaps I'm missing something). You could try making a pyvips mask that's wider than you need and removing the first and last columns. But you'll probably find it makes little difference.

**Reply by jcupitt** [reply]
Also, I've not been involved with govips, I don't know how well it works. I would make a quick prototype system and test it for reliability, memory behaviour, speed, threading behaviour, performance under very heavy load, etc. before committing to it. Some of the third party libvips bindings have serious issues (but maybe govips is OK? I don't know).

**Reply by Ruchika-K** [reply]
Okay . Got it. Thanks. Will verify if this mask generated is within tolerance. 
Regarding using govips, we had done a short poc. Although proper perf testing is yet to be done.
